node/node.go

```diff
@@ -363,66 +367,277 @@ func (node *Node) AddPendingReceipts(receipts *types.CXReceiptsProof) {
+	errChan := make(chan withError)
+
+	type p2pHandlerConsensus func(
+		ctx context.Context,
+		msg *msg_pb.Message,
+	) error
 
-	for i, topic := range allTopics {
-		sub, err := topic.Subscribe()
+	type p2pHandlerElse func(
+		ctx context.Context,
+		rlpPayload []byte,
+	) error
+
+	type validated struct {
+		consensusBound bool
+		handleC        p2pHandlerConsensus
+		handleCArg     *msg_pb.Message
+		handleE        p2pHandlerElse
+		handleEArg     []byte
+	}
+
+	isThisNodeAnExplorerNode := node.NodeConfig.Role() == nodeconfig.ExplorerNode
+
+	for i := range allTopics {
+		sub, err := allTopics[i].Topic.Subscribe()
 		if err != nil {
 			return err
 		}
-		weighted[i] = semaphore.NewWeighted(maxMessageHandlers)
-		msgChan := make(chan *libp2p_pubsub.Message)
 
-		go func(msgChan chan *libp2p_pubsub.Message, sem *semaphore.Weighted) {
-			for msg := range msgChan {
-				payload := msg.GetData()
-				if len(payload) < p2pMsgPrefixSize {
-					continue
+		topicNamed := allTopics[i].Name
+		isConsensusBound := allTopics[i].consensusBound
+
+		utils.Logger().Info().
+			Str("topic", topicNamed).
+			Msg("enabled topic validation on consensus bound messages")
+
+		if err := pubsub.RegisterTopicValidator(
+			topicNamed,
+			func(ctx context.Context, peer libp2p_peer.ID, msg *libp2p_pubsub.Message) bool {
+				hmyMsg := msg.GetData()
+				if len(hmyMsg) < p2pMsgPrefixSize {
+					errChan <- withError{
+						errors.WithStack(errors.Wrapf(
+							errMsgHadNoHMYPayLoadAssumption, "on topic %s", topicNamed,
+						)), nil,
+					}
+					return false
 				}
-				if sem.TryAcquire(1) {
-					go func() {
-						node.HandleMessage(
-							payload[p2pMsgPrefixSize:], msg.GetFrom(),
-						)
-						sem.Release(1)
-					}()
-				} else {
-					utils.Logger().Warn().
-						Msg("could not acquire semaphore to process incoming message")
+
+				openBox := hmyMsg[p2pMsgPrefixSize:]
+
+				switch proto.MessageCategory(openBox[proto.MessageCategoryBytes-1]) {
+				case proto.Consensus:
+					if !isConsensusBound {
+						errChan <- withError{
+							errors.WithStack(errConsensusMessageOnUnexpectedTopic), msg,
+						}
+						return false
+					}
+
+					validMsg, err := validateShardBoundMessage(
+						context.TODO(), openBox[proto.MessageCategoryBytes:],
+					)
+
+					if err != nil {
+						errChan <- withError{
+							errors.WithStack(errMsgHadNoHMYPayLoadAssumption), msg,
+						}
+						return false
+					}
+					msg.ValidatorData = validated{
+						consensusBound: true,
+						handleC:        node.Consensus.HandleMessageUpdate,
+						handleCArg:     validMsg,
+					}
+					return true
+
+				case proto.Node:
+					// TODO push the message parsing here, so can ban
+					msg.ValidatorData = validated{
+						consensusBound: false,
+						handleE:        node.HandleNodeMessage,
+						handleEArg:     openBox,
+					}
+				default:
+					return false
 				}
+
+				return true
+
+			},
+			libp2p_pubsub.WithValidatorTimeout(24),
```

Comment:
24 seems not reasonable. It's Nano seconds.

Reply:
agree, set to 10 millisecond, so timeout on 100 messages per second.

Reply:
validation should be very fast, only check a few byte. 100 nanoseconds should be enough.